### ISSUE TITLE
CI fails to start container frequently

### ISSUE THREAD

**Issue by cofyc** [pingcap/tidb-operator#1603]
## Bug Report

example logs: https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/operator_ghpr_e2e_test_kind/detail/operator_ghpr_e2e_test_kind/2484/pipeline/72

example failures:

```
[2020-01-27T14:16:18.322Z] docker run error: command "docker run --hostname tidb-operator-worker3 --name tidb-operator-worker3 --label io.x-k8s.kind.role=worker --privileged --security-opt seccomp=unconfined --security-opt apparmor=unconfined --tmpfs /tmp --tmpfs /run --volume /var --volume /lib/modules:/lib/modules:ro --detach --tty --label io.x-k8s.kind.cluster=tidb-operator kindest/node:v1.12.10@sha256:68a6581f64b54994b824708286fafc37f1227b7b54cbb8865182ce1e036ed1cc" failed with error: exit status 125

[2020-01-27T14:16:19.690Z] ERROR: failed to create cluster: docker run error: command "docker run --hostname tidb-operator-worker3 --name tidb-operator-worker3 --label io.x-k8s.kind.role=worker --privileged --security-opt seccomp=unconfined --security-opt apparmor=unconfined --tmpfs /tmp --tmpfs /run --volume /var --volume /lib/modules:/lib/modules:ro --detach --tty --label io.x-k8s.kind.cluster=tidb-operator kindest/node:v1.12.10@sha256:68a6581f64b54994b824708286fafc37f1227b7b54cbb8865182ce1e036ed1cc" failed with error: exit status 125

[2020-01-27T14:16:19.690Z] 

[2020-01-27T14:16:19.690Z] Output:

[2020-01-27T14:16:19.690Z] f57dba5be9f8b69d9f6fc887d7a12ba5032ea26fb452a8f544ea64adea31e1f0

[2020-01-27T14:16:19.691Z] docker: Error response from daemon: OCI runtime create failed: container_linux.go:346: starting container process caused "process_linux.go:297: applying cgroup configuration for process caused \"mkdir /sys/fs/cgroup/memory/docker/f57dba5be9f8b69d9f6fc887d7a12ba5032ea26fb452a8f544ea64adea31e1f0: cannot allocate memory\"": unknown.

[2020-01-27T14:16:19.691Z] 

[2020-01-27T14:16:19.691Z] Stack Trace: 

[2020-01-27T14:16:19.691Z] sigs.k8s.io/kind/pkg/errors.WithStack

[2020-01-27T14:16:19.691Z] 	/src/pkg/errors/errors.go:51

[2020-01-27T14:16:19.691Z] sigs.k8s.io/kind/pkg/exec.(*LocalCmd).Run

[2020-01-27T14:16:19.691Z] 	/src/pkg/exec/local.go:116

[2020-01-27T14:16:19.691Z] sigs.k8s.io/kind/pkg/cluster/internal/providers/docker.createContainer

[2020-01-27T14:16:19.691Z] 	/src/pkg/cluster/internal/providers/docker/provision.go:99

[2020-01-27T14:16:19.691Z] sigs.k8s.io/kind/pkg/cluster/internal/providers/docker.planCreation.func3

[2020-01-27T14:16:19.691Z] 	/src/pkg/cluster/internal/providers/docker/provision.go:89

[2020-01-27T14:16:19.691Z] sigs.k8s.io/kind/pkg/errors.UntilErrorConcurrent.func1

[2020-01-27T14:16:19.691Z] 	/src/pkg/errors/concurrent.go:30

[2020-01-27T14:16:19.691Z] runtime.goexit

[2020-01-27T14:16:19.691Z] 	/usr/local/go/src/runtime/asm_amd64.s:1357
```

example logs from `kubectl describe pods`:

```
Warning  FailedCreatePodContainer  14m (x10 over 16m)  kubelet, 172.16.5.64  unable to ensure pod container exists: failed to create container for [kubepods podd1e2a265-410f-11ea-b31c-d0946604d177] : mkdir /sys/fs/cgroup/memory/kubepods/podd1e2a265-410f-11ea-b31c-d0946604d177: cannot allocate memory
```

related issue:

- https://github.com/docker/for-linux/issues/841

possible solutions:

- upgrade kernel to `kernel-3.10.0-1062.4.1`+  (https://github.com/docker/for-linux/issues/841#issuecomment-573518203)

**Comment by cofyc** [comment]
logs of kubelet:

```
...
Jan 29 14:14:16 tidb-operator-control-plane kubelet[2082]: I0129 14:14:16.196046    2082 cpu_manager.go:173] [cpumanager] starting with none policy
Jan 29 14:14:16 tidb-operator-control-plane kubelet[2082]: I0129 14:14:16.196057    2082 cpu_manager.go:174] [cpumanager] reconciling every 10s
Jan 29 14:14:16 tidb-operator-control-plane kubelet[2082]: I0129 14:14:16.196067    2082 policy_none.go:43] [cpumanager] none policy: Start
Jan 29 14:14:16 tidb-operator-control-plane kubelet[2082]: E0129 14:14:16.197921    2082 node_container_manager_linux.go:50] Failed to create ["kubepods"] cgroup
Jan 29 14:14:16 tidb-operator-control-plane kubelet[2082]: F0129 14:14:16.197938    2082 kubelet.go:1380] Failed to start ContainerManager mkdir /sys/fs/cgroup/memory/kubepods: cannot allocate memory
Jan 29 14:14:16 tidb-operator-control-plane systemd[1]: kubelet.service: Main process exited, code=exited, status=255/EXCEPTION
Jan 29 14:14:16 tidb-operator-control-plane systemd[1]: kubelet.service: Failed with result 'exit-code'.
```

**Comment by cofyc** [comment]
It still happens on the new kernel version (3.10.0-1062.9.1.el7.x86_64). It's possible that we're leaking memory cgroups.

```
# cat !$
cat /proc/cgroups
#subsys_name	hierarchy	num_cgroups	enabled
cpuset	3	3201	1
cpu	5	6568	1
cpuacct	5	6568	1
memory	7	8331	1
devices	4	8616	1
freezer	2	2693	1
net_cls	8	2708	1
blkio	11	7913	1
perf_event	10	3046	1
hugetlb	6	2947	1
pids	9	8812	1
net_prio	8	2708	1
```

There are lots of empty cgroups (no process in the cgroup) on the machine.

**Comment by cofyc** [comment]
however, the job pod is terminated with 0  second grace period. We should find out why...

```
Feb 06 12:28:34 172.16.5.67 kubelet[15125]: I0206 12:28:34.209179   15125 kubelet.go:1899] SyncLoop (DELETE, "api"): "build-tidb-operator-master-kind-249-9b2v8-gz33b-2krwn_jenkins-ci(d654cf7e-4893-11ea-b4a3-1866da9d20ee)"
Feb 06 12:28:34 172.16.5.67 kubelet[15125]: I0206 12:28:34.209712   15125 kuberuntime_container.go:555] Killing container "docker://280bf48b2cda6f0dcc2bb4a6ac317913962054d70a00ed974bfe280df78b1719" with 0 second grace period
Feb 06 12:28:34 172.16.5.67 kubelet[15125]: I0206 12:28:34.209802   15125 kuberuntime_container.go:555] Killing container "docker://b38a77770d867e24ecd606fc4a9391763c35553b61de35f7f8d013d49c7c6a0a" with 0 second grace period
Feb 06 12:28:34 172.16.5.67 kubelet[15125]: I0206 12:28:34.211965   15125 kubelet.go:1893] SyncLoop (REMOVE, "api"): "build-tidb-operator-master-kind-249-9b2v8-gz33b-2krwn_jenkins-ci(d654cf7e-4893-11ea-b4a3-1866da9d20ee)"
Feb 06 12:28:34 172.16.5.67 kubelet[15125]: I0206 12:28:34.234719   15125 kubelet_pods.go:1083] Killing unwanted pod "build-tidb-operator-master-kind-249-9b2v8-gz33b-2krwn"
Feb 06 12:28:34 172.16.5.67 kubelet[15125]: I0206 12:28:34.234785   15125 kuberuntime_container.go:555] Killing container "docker://280bf48b2cda6f0dcc2bb4a6ac317913962054d70a00ed974bfe280df78b1719" with 0 second grace period
Feb 06 12:28:34 172.16.5.67 kubelet[15125]: I0206 12:28:34.234794   15125 kuberuntime_container.go:555] Killing container "docker://b38a77770d867e24ecd606fc4a9391763c35553b61de35f7f8d013d49c7c6a0a" with 0 second grace period
```

**Comment by cofyc** [comment]
`pod.Spec.TerminationGracePeriodSeconds` is 30 seconds, so `pod.DeletionGracePeriodSeconds` is set to 0 in deletion. https://github.com/kubernetes/kubernetes/blob/641321c94c3e0943a57230596164e2e8f2eb2c52/pkg/kubelet/kuberuntime/kuberuntime_container.go#L568-L575

**Comment by cofyc** [comment]
I reported an issue for Jenkins Kubernetes Plugin: https://issues.jenkins-ci.org/browse/JENKINS-60987. Right now, we need to recycle leaked resources periodically.

**Comment by cofyc** [comment]
*Update*: even if all empty cgroups are destoryed, some machines still fails "Cannot allocate memory" error. I guess there is internal cgroup or memory leak in kernel ( 3.10.0-957.el7.x86_64, 3.10.0-1062.9.1.el7.x86_64 )

**Comment by cofyc** [comment]
Updated: `"SLUB: Unable to allocate memory on node -1"` issue mentioned in https://pingcap.com/blog/try-to-fix-two-linux-kernel-bugs-while-testing-tidb-operator-in-k8s/#bug-1-unstable-kmem-accounting had been fixed in `1062.9.1.el7.x86_64`. But the memory cgroup leaking still exists. Addin `cgroup.memory=nokmem` kernel parameter will work around this issue.
